crates/sui-rosetta/src/operations.rs

```diff
@@ -558,6 +558,74 @@ impl Operations {
         };
         balance_change.chain(gas)
     }
+
+    fn process_single_gascoin_transfer(
+        coin_change_operations: &mut impl Iterator<Item = crate::operations::Operation>,
+        tx: SuiTransactionBlockKind,
+        sender: SuiAddress,
+        gas_used: i128,
+    ) -> Vec<Operation> {
+        let mut is_single_gascoin_transfer = false;
+        match tx {
+            SuiTransactionBlockKind::ProgrammableTransaction(pt) => {
+                let SuiProgrammableTransactionBlock {
+                    inputs: _,
+                    commands,
+                } = &pt;
+                for command in commands {
+                    match command {
+                        SuiCommand::TransferObjects(objs, _) => {
+                            if objs.len() == 1 {
+                                match objs[0] {
+                                    SuiArgument::GasCoin => {
+                                        is_single_gascoin_transfer = true;
+                                    }
+                                    _ => {}
+                                }
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            _ => {}
+        }
+        if !is_single_gascoin_transfer {
+            return vec![];
+        };
+        let mut operations = vec![];
+        coin_change_operations.into_iter().for_each(|operation| {
+            match operation.type_ {
+                OperationType::Gas => {
+                    // change gas account back to sender as the sender is the one
+                    // who paid for the txn (this is the format Rosetta wants to process)
+                    operations.push(Operation::gas(sender, gas_used))
+                }
+                OperationType::SuiBalanceChange => {
+                    operation.account.map(|account| {
+                        let mut amount = match operation.amount {
+                            Some(amount) => amount.value,
+                            None => 0,
+                        };
+                        if account.address == sender {
+                            // sender's balance needs to be adjusted for gas
+                            amount -= gas_used;
+                        } else {
```

Comment:
We should not edit every balance-change. Only the gas-coin sender and the recipient ones.

Reply:
And we should also make sure that these additions/subtractions each happen only once.

Reply:
Done